### DISCUSSION TITLE
How to generate pore network with different connectivity?

### DISCUSSION THREAD

**Discussion by mourendong** [PMEAL/OpenPNM#2370] [Q&A] [answered]
Hi, everyone,
I want to generate pore networks with various connectivity numbers. The "op. network. Cubic" only generate the network with the connectivity number of four. 
As the figure shows, the first one with a connectivity number of three, and the second one with a connectivity number of six. Can you help to how to generate the network like this style?
Thanks for your kind help!
![image](https://user-images.githubusercontent.com/47968460/163710839-7e826352-3b36-4d40-b44e-5b14a5dc5f77.png)

**Comment by Devin-one** [comment]
You can take a look at this website. https://github.com/PMEAL/OpenPNM/blob/dev/examples/tutorials/network/cubic_networks.ipynb
 I hope it will help you. By the way, I would like to ask if there is any tutorial for calculating porosity and generating related images.

**Comment by jgostick** [comment, marked as answer]
Another options is to put points in 2D or 3D space, then pass those to the Voronoi class.  For instance, if you generate a triangular lattice of points using PoreSpy (see [example here](https://porespy.org/examples/generators/reference/lattice_spheres.html#lattice)), with a radius of 1 to get points instead of spheres, then you can use ``pts = np.where(~im)`` to get the x,y locations of each point (in voxels).  Passing these into the ``Voronoi`` class *should* generate the hexahedral grid above.  It also occurs to me that if you use the DelaunayVoronoiDual class you will get the triangular mesh as the combination of delaunay and voronoi networks, plus the interconnections.  I haven't tried this myself, but I'm 99% sure it should just work.

**Reply by jgostick** [reply]
I should add that if you want to do 2D you'll need to add a column of z=0's to the x,y values returned from ``np.where`` since the OpenPNM classes will want that.

**Reply by mourendong** [reply]
Thanks for your kind help! I have generated the network through your suggestion.

**Reply by jgostick** [reply]
great!  Can you post a picture of your result?  Maybe share a code snippet as well, for future users?

**Reply by mourendong** [reply]
Okay! I generate one N×N to list pores coordinate. For the pores in the first and second columns, the coordinate is generated one by one. Then, based on this, the pores in other columns were generated. Finally, the network with triangle distribution was generated by the 'Delaunay' class. The way is cumbersome, bus useful.
Thank you again for your advice.
![image](https://user-images.githubusercontent.com/47968460/164356103-03194515-1630-4a92-8ab7-8066f674da1b.png)
``` python
import numpy as np
import openpnm as op
import matplotlib.pyplot as plt
N=12
num=[[0 for x in range (0,2)] for y in range (0,N**2)]
for x in range (0, N**2): 
    if x in range(1,N):
        num[x][1]=2*x-1
    if x in range(N+1,2*N-1):
        num[x][0]=3**0.5
        num[x][1]=2*(x-N)
    if x ==N:
        num[x][0]=3**0.5
    if x ==2*N-1:
        num[x][1]=2*N-3
        num[x][0]=3**0.5 
for x in range(0,N):
    if x ==0:
        for y in range(0,N**2):
            if y%N==0 and y!=0: 
                num[y][0]=y/N*3**0.5
    if x!=0:
        for y in range(0,N**2):
            if (y-x)%(2*N)==0 and y>x:
                num[y][0]=(y-x)/(2*N)*2*3**0.5
                num[y][1]=num[x][1]
for x in range(N,2*N):
    for y in range(0,N**2):
        if (y-x)%(2*N)==0 and y>x:
            num[y][0]=(2*(y-x)/(2*N)+1)*3**0.5
            num[y][1]=num[x][1]                    
pn = op.network.Delaunay(points=num, shape=[500, 500, 0])
print(pn)
fig=plt.figure()
op.topotools.plot_connections(network=pn, throats=pn.throats(['all']), c='g', linewidth=0.5,fig=fig)
_ = plt.axis('off')
```
